netcore/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -24,6 +24,10 @@
 using Internal.Runtime.Augments;
 using Internal.Threading.Tasks;
 
+#if !CORECLR
```

Comment:
This needs to be `#if CORERT`

Reply:
I was not sure about it. Certainly depends on how the code is shared and whether you consider `System.Threading.Thread` or `Internal.Runtime.Augments.RuntimeThread` the main class. In CoreCLR it's mixed together in a not so nice way, in CoreRT `System.Threading.Thread` doesn't exist at all. There are some dirty tricks for other classes, where `using Thread = ...` at the beginning of the shared file is ignored because the default type resolution prefers to use the `Thread` from the current namespace (which happens to be System.Threading[.*]).

I have two branches for CoreCLR where I try to clean it up using different approaches.

The less invasive approach is to remove all `using Thread = ...` lines and strictly use `RuntimeThread` everywhere in the code. The `System.Threading.Thread` class still exists, but it's only used by unmanaged code, `RuntimeThread` itself and possibly a debugger.

The more invasive approach is to remove `System.Threading.Thread` from CoreLib altogether and move all the code to `RuntimeThread`. It is cleaner in many places, but the comments suggest that it will likely break Visual Studio debugger.

Reply:
For reference, here are links to the branches:

- Conservative approach (using RuntimeThread everywhere, but keeping the implementation in Thread): https://github.com/filipnavara/coreclr/tree/runtimethread2
- Invasive approach (removing Thread altogether in favor of RuntimeThread): https://github.com/filipnavara/coreclr/tree/runtimethread

If any of that approach makes sense I can work on it further, but I definitely don't have enough insights on whether it's the right direction at all. If you just want to flip the `#if` and be done with it I don't really mind. /cc @jkotas

Reply:
You can follow the rule that CoreRT is the odd one (it's not always true but very common)

Reply:
@marek-safar Generally I do. In this specific instance I had a feeling that CoreRT was already cleaned-up and CoreCLR was not (for various historical reasons). It somehow feels wrong to have two unrelated `System.Threading.Thread` types in one application.

Reply:
The third option is to move System.Thread from CoreFX back to CoreLib, including its transitive dependency closure. We should either do that, or remove Thread altogether in CoreLib in favor of RuntimeThread (also change RuntimeThread namespace to Internal.Threading.RuntimeThread to make it match the convention we are using for most types like that). Having two System.Threading.Thread types in the system is just a source of problem (even if one of them is internal).

@marek-safar We have touched on this during our call last week. Do you have an opinion about the approach you would prefer for this?

Reply:
I think the third option may be the easiest one for sharing with Mono. However the transient closure is not very "core" anymore (`IPrincipal` and its dependencies). I have submitted PR for removing Thread in CoreLib and it passes CI, but there is unsolved issue with VS debugger (the comments in code could be outdated, but I didn't give it a try yet). And if it were to make the sharing harder I am fine with abandoning it in favor of a different approach. I'll try to see what happens with @marek-safar's Mono branch in combination with my PR.

Reply:
I agree that having two types (Thread and RuntimeThread) is quite confusing and my prefered solution would be to standardize on one internal type in CoreLib with partial parts for runtime specific bits as we do elsewhere.

@jkotas the Thread is quite API complex and although it could be moved to CoreLib fully unless we need ExecutionContext and AppDomain in CoreLib RuntimeThread could work fine.

Reply:
ExecutionContext is needed in CoreLib (e.g. async methods depend on it). AppDomain is not needed in CoreCLR CoreLib.